File: ueueo-framework/ueueo-eventbus/src/main/java/com/ueueo/eventbus/distributed/IDistributedEventBus.java
package com.ueueo.eventbus.distributed;

import com.ueueo.exception.SystemException;
import com.ueueo.disposable.IDisposable;
import com.ueueo.eventbus.IEventBus;

/**
 * @author Lee
 * @date 2022-05-20 16:41
 */
public interface IDistributedEventBus extends IEventBus {
    /**
     * Registers to an event.
     * Same (given) instance of the handler is used for all event occurrences.
     *
     * @param eventType Event type
     * @param handler   Object to handle the event
     *
     * @return
     */
    IDisposable subscribe(Class<?> eventType, IDistributedEventHandler handler);

    /**
     * @param eventType
     * @param eventData
     * @param onUnitOfWorkComplete true
     * @param useOutbox            true
     */
    void publish(Class<?> eventType, Object eventData, Boolean onUnitOfWorkComplete, Boolean useOutbox);

    default ISupportsEventBoxes asSupportsEventBoxes() {
        if (!(this instanceof ISupportsEventBoxes)) {
            throw new SystemException("Given type ({eventBus.GetType().AssemblyQualifiedName}) should implement {nameof(ISupportsEventBoxes)}!");
        }
        return (ISupportsEventBoxes) this;
    }
}

File: ueueo-framework/ueueo-eventbus/src/main/java/com/ueueo/eventbus/distributed/IDistributedEventBus.cs
using System;
using System.Threading.Tasks;

namespace Volo.Abp.EventBus.Distributed;

public interface IDistributedEventBus : IEventBus
{
    /**
     * Registers to an event.
     * Same (given) instance of the handler is used for all event occurrences.
    *
     * <typeparam name="TEvent">Event type</typeparam>
     * <param name="handler">Object to handle the event</param>
     */
    IDisposable Subscribe<TEvent>(IDistributedEventHandler<TEvent> handler)
        //where TEvent : class;

    void PublishAsync<TEvent>(
        TEvent eventData,
        boolean onUnitOfWorkComplete = true,
        boolean useOutbox = true)
        //where TEvent : class
        ;

    void PublishAsync(
        Type eventType,
        Object eventData,
        boolean onUnitOfWorkComplete = true,
        boolean useOutbox = true);
}
